MiniTwitch.Common/AsyncEventCoordinator.cs

```diff
@@ -0,0 +1,138 @@
+﻿namespace MiniTwitch.Common;
+
+/// <summary>
+/// Provides mechanisms to asynchronously wait for an event to occur
+/// </summary>
+/// <typeparam name="TEnum">The enum type to operate on</typeparam>
+public class AsyncEventCoordinator<TEnum> : IDisposable
+    where TEnum : struct
+{
+    private readonly Dictionary<int, SemaphoreSlim> _locks;
+
+    /// <summary>
+    /// Initializes a new instance of AsyncEventCoordinator
+    /// </summary>
+    /// <param name="initialCount">The initial semaphore request count</param>
+    /// <exception cref="NotSupportedException"></exception>
+    public AsyncEventCoordinator(int initialCount = 0)
+    {
+        if (!typeof(TEnum).IsEnum)
+            throw new NotSupportedException($"The type {typeof(TEnum)} is not an enum");
+
+        TEnum[] values = (TEnum[])Enum.GetValues(typeof(TEnum));
+        _locks = new(values.Length);
+        for (int i = 0; i < values.Length; i++)
+        {
+            _locks[(int)(object)values[i]] = new SemaphoreSlim(initialCount);
+        }
```

Comment:
convert to `FrozenDictionary` at the end

Reply:
net8+ only, so can't do that